### PR TITLE
closes #149: Implement MapQuery endpoint search

### LINKED ISSUE #149: Make MapQuery actually search
Currently, the MapQuery endpoint is still hard-coding the programs that come back.  Make that work correctly so we can demo functionality for Elise.

### PR DESCRIPTION
**Pull request by RyanBard**
* Implements searching by County, ProgramArea, Organization, and Program for
  the KML url to give google.
* Changes KmlService to draw purple filled shapes for counties in the program
  results and gray filled shapes for all other counties.
* Adds @Nonnull annotations and requireNonNull calls.
* Adds ordering by name to getAll-Programs.
* Adds ordering by name to getAll-Organizations.
* Adds comment to sql files to disable IntelliJ warnings about datasources.
* Comments out counties with odd geometry that were causing problems in the
  KML generation code (this will be fixed in a follow up PR).
* Adjusts county associations so they are more sensible.
* Updates unit tests.

### CODE REVIEW

File: src/main/java/org/hmhb/organization/DefaultOrganizationService.java [hunk 1]
```diff
@@ -59,9 +58,7 @@ public Organization getById(
     @Override
     public List<Organization> getAll() {
         LOGGER.debug("getAll called");
-        List<Organization> target = new ArrayList<>();
-        dao.findAll().forEach(target::add);
-        return target;
+        return dao.findAllByOrderByNameAsc();
```
Comment (RyanBard) [inline src/main/java/org/hmhb/organization/DefaultOrganizationService.java]: I added this so organizations would be ordered and it reduced the code a little bit.

File: src/main/java/org/hmhb/program/DefaultProgramService.java [hunk 2]
```diff
@@ -69,16 +74,169 @@ public Program getById(
...
     }
 
     @Timed
     @Override
     public List<Program> getAll() {
         LOGGER.debug("getAll called");
-        List<Program> target = new ArrayList<>();
-        dao.findAll().forEach(target::add);
-        return target;
+        return dao.findAllByOrderByNameAsc();
```
Comment (RyanBard) [inline src/main/java/org/hmhb/program/DefaultProgramService.java]: Same here: programs will be ordered and the code is reduced.

File: src/main/java/org/hmhb/program/DefaultProgramService.java [hunk 3]
```diff
@@ -69,16 +74,169 @@ public Program getById(
...
+
+        /*
+         * Both of these potential implementations (nested ifs and flattened ifs) are really ugly,
+         * but using JPA Specification looks much more complicated.
+         *
+         * For now, with only 3 search params, I'm just going to have all permutations of
+         * searches in the DAO.  If anymore params are added and the permutations get worse,
+         * I'll have to use JPA Specification or figure something else out.
+         */
+//        return doSearchFlattenedIfs(organizationId, countyId, programAreaId);
```
Comment (RyanBard) [inline src/main/java/org/hmhb/program/DefaultProgramService.java]: I don't really like either of these implementations.  @mckoon please review them and give a preference on which to keep.

The alternative (JPA Specification) seemed much worse for our needs (at least at the moment):

https://www.petrikainulainen.net/programming/spring-framework/spring-data-jpa-tutorial-part-four-jpa-criteria-queries/
http://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/domain/Specifications.html

There was something mentioned called Querydsl that I didn't look into much, but I didn't want to drag in another tech/library:

https://spring.io/blog/2011/04/26/advanced-spring-data-jpa-specifications-and-querydsl/

File: src/main/java/org/hmhb/program/DefaultProgramService.java [hunk 4]
```diff
@@ -69,16 +74,169 @@ public Program getById(
     ) {
         LOGGER.debug("getByIds called: ids={}", ids);
         requireNonNull(ids, "ids cannot be null");
-        return dao.findByIdIn(ids);
+        return dao.findAll(ids);
```
Comment (RyanBard) [inline src/main/java/org/hmhb/program/DefaultProgramService.java]: It turns out that JpaRepository already had a bulk lookup method, so I took out our findByIdIn and used that.

File: src/main/resources/data.sql [hunk 5]
```diff
@@ -1,3 +1,5 @@
+-- noinspection SqlNoDataSourceInspectionForFile
```
Comment (RyanBard) [inline src/main/resources/data.sql]: IntelliJ was being annoying with inappropriate warnings in our sql files.

File: src/main/resources/data.sql [hunk 6]
```diff
@@ -335,15 +337,15 @@ INSERT INTO county (
...
-);
+-- INSERT INTO county (
```
Comment (RyanBard) [inline src/main/resources/data.sql]: I'm going to follow up with a fix for these counties later.

File: src/main/resources/data.sql [hunk 7]
```diff
@@ -1808,39 +1810,39 @@ INSERT INTO program_county (
     county_id
 ) VALUES (
     1,
-    1
+    30
```
Comment (RyanBard) [inline src/main/resources/data.sql]: I changed the counties covered to be at least the county they are located in.

File: src/main/java/org/hmhb/program/DefaultProgramService.java (hunk 3 above)
Comment (mckoon) [inline src/main/java/org/hmhb/program/DefaultProgramService.java]: I went ahead and merged this in last night with both implementations in here, and using the nested one.
Now they are in here part of history, so if we decide, they can be recovered.

Based on the code here and the linked articles, I think creating a `Specification` for each and using them in here would probably be the cleanest.  I agree that we probably don't want to drag in Querydsl.

Something like the below pseudo code would scale if more criteria are added, and reads cleaner IMO:

```
        Specifications specifications = where(ALWAYS_TRUE);

        if (organizationId != null) {
            specifications = specifications.and(
                    hasOrganizationId(organizationId)
            );
        }

        if (countyId != null) {
            specifications = specifications.and(
                    hasCountyId(countyId)
            );
        }

        if (programAreaId != null) {
            specifications = specifications.and(
                    hasProgramAreaId(programAreaId)
            );
        }

        return dao.findAll(specifications);
```

**Pull request merged**